### DISCUSSION TITLE
Serverside disconnects

### DISCUSSION THREAD

**Discussion by Richard-Whalley** [SteamRE/SteamKit#1448] [General]
Hi, I've been building a PoC using steamkit2, and so far the experience has been great as its more feature complete than the rust library I was originally using; however today, I've started to experience serverside disconnects, so as soon as I call client.connect(), I get an onDisconnect callback, and the disconnect is initiated from the steam side; It happened suddenly. I've tried other wifi/connections just in case it was geolocation based, but same issues. There are other people in my discord community that are experiencing the same issues.

**Comment by Shadowtail117** [comment]
I have been experiencing the same thing. My bot is designed to fetch lobby information from a couple of games and has worked fine for the past year and a half, but ever since last Friday, the bot has been completely unable to log in. Every attempt is met with an immediate OnDisconnected callback. At first, I thought it was because my version of SteamKit was outdated, but on the newest version, the same thing happened. This is happening to both my remotely hosted production bot and my locally hosted debug bot, so I don't think it's an issue of connection with Steam servers.

If anyone has some ideas on how to fix it, it'd be greatly appreciated.

**Comment by xPaw** [comment]
I suspect there's some issues happening with Steam servers, because mobile chat app also stopped working last week: https://steamcommunity.com/discussions/forum/8/4849905062599698735/

There are issues like GetCMList giving servers on the opposite side of the world, so I can't really imagine why it would be an issue with SteamKit.

**Reply by Richard-Whalley** [reply]
Yep, so the main issue I wasn't doing retries on server disconnect, it seems to take a couple of attempts to find a server that works, thankfully this will make my code more robust in the the long term, thanks for the speedy reply.

**Comment by Richard-Whalley** [comment]
```
void OnDisconnected(SteamClient.DisconnectedCallback callback)
{
    if (callback.UserInitiated != false) return;
    Console.WriteLine("Steam terminated unexpectedly, attempting to reconnect...");
    Connect();
}
```
Something along the lines of the above. @Shadowtail117

**Reply by Shadowtail117** [reply]
I tried this out but the bot was never able to connect even after a few dozen attempts. How long did it take you to successfully log in?

**Reply by Richard-Whalley** [reply]
pretty reliably in 3-6 attempts, so a 100ms or so

**Reply by Shadowtail117** [reply]
I had my bot set to try reconnecting every 5 seconds which never seemed to work, but when I changed it to try every second, it managed to get in within a few tries. Not sure why that made all the difference, but in either case I'm glad it's finally working again. Thank you for the help!

EDIT: It's still fairly inconsistent, I guess I just got lucky before. Hoping Steam fixes whatever happened quickly.

**Comment by xPaw** [comment]
There appears to be an issue that some CM servers only support TLS 1.3, like cmp1-vie1.steamserver.net:443 for example. I can connect just fine to it on Windows 11, but someone else fails to on Windows 10.

```
$curl -v --tlsv1.2 --tls-max 1.2 https://cmp1-vie1.steamserver.net:443
curl: (35) OpenSSL/3.0.14: error:0A00042E:SSL routines::tlsv1 alert protocol version
```

Can you guys test that theory?

**Reply by kuznetsov-kd** [reply]
I solve the problem on WIndows 10 following this guide https://answers.microsoft.com/en-us/windows/forum/all/windows-10-tls-13-enablement-registry-keys/963d0ff4-9b07-4102-9685-e32d1d25f8cc
It didn't worked and I tried to delete HKLM\SYSTEM\CurrentControlSet\Control\SecurityProviders\SCHANNEL\Protocols\TLS 1.2\ section.
And unexpectedly it worked.

I'm not sure it's a good solution but in some cases could be.

**Reply by yaakov-h** [reply]
I can reach that CM with that command line from curl on macOS without problems.

> \* TLSv1.2 (IN), TLS handshake, Finished (20):
> \* SSL connection using TLSv1.2 / ECDHE-ECDSA-CHACHA20-POLY1305 / [blank] / UNDEF
> \* ALPN: server accepted http/1.1

**Reply by xPaw** [reply]
I can too now, maybe they fixed it (I did tell someone at Valve).

**Reply by kuznetsov-kd** [reply]
Oh, yep. I was wrong.
According to the ssllabs they enabled TLS 1.2 again
https://www.ssllabs.com/ssltest/analyze.html?d=cmp2-vie1.steamserver.net

**Reply by Shadowtail117** [reply]
Was able to reach that CM on my Windows 10 as well. I haven't messed with CM configuration with my program at all and have been using the default, do you think having a specific configuration would help fix the issue of not being able to connect? I haven't been able to connect using Steamkit at all in the past couple days no matter what (attached video is pretty much the only thing I've been getting; I set it to try to connect 10x/second and still nothing even if I let it run for a full minute).

https://github.com/user-attachments/assets/f7c19058-74d7-4a2b-b6e1-b61b67877e93

**Comment by tpill90** [comment]
I’m getting the same issue on SteamPrefill, and I’m getting a lot of reports on this issue : [issues/376](https://github.com/tpill90/steam-lancache-prefill/issues/376). 

I haven’t been able to put in much time to look at it, but it seems that some regions are less affected than others. And it’s possibly a rolling issue as it seems to come and go. 

It seems to me that this has been around for a month or so, around the time that I submitted #1420.

**Reply by tpill90** [reply]
I updated to the latest 3.0.0-beta5 build and had some users test across various regions, and it appears to be reliably connecting again.  I've tested various regions myself that weren't covered by other users and those are working as well.  Thanks for the work on figuring this out @xPaw

**Comment by ChrisN944** [comment]
Glad to see this is an issue with Steam and not my code just randomly imploding... my sanity has been restored.

I don't entirely understand the discussion regarding CM as I have limited experience with SteamKit2. I am blocked by this issue and can't continue developing my project until I have a work-around/ the issue is solved. Is the consensus that this is an issue that Steam needs to resolve (should they decide to resolve it)?

**Reply by tpill90** [reply]
The latest beta of SteamKit (beta5) looks to have resolved this issue in my app.  If you want to see a real world example of how I'm interacting with SteamKit then check this file out [Steam3Session.cs](https://github.com/tpill90/steam-lancache-prefill/blob/master/SteamPrefill/Handlers/Steam/Steam3Session.cs).  @Shadowtail117 You might also find this helpful

**Comment by Richard-Whalley** [comment]
It’s always worth handling server side OnDisconnect callbacks to re-establish the connection correctly imo. With a proper retry strategy in place this works in its current state. Steam are fixing this, and I know some regions have achieve stability. I’m EU west, and it takes around 3 attempts on average to get a connection.On Oct 18, 2024, at 7:42 AM, ChrisN944 ***@***.***> wrote:﻿
Glad to see this is an issue with Steam and not my code just randomly imploding... my sanity has been restored.
I don't entirely understand the discussion regarding CM as I have limited experience with SteamKit2. I am blocked by this issue and can't continue developing my project until I have a work-around/ the issue is solved. Is the consensus that this is an issue that Steam needs to resolve (should they decide to resolve it)?

**Reply by ChrisN944** [reply]
Unfortunately, that was the very first I did thing when I started my project. It doesn't work. If I try to connect and I cannot I retry 3 times. If that fails I move on to the next account. Every time I retry I fail to connect due to a Steam-initiated disconnect. I cannot get any of my accounts to connect. I'm US East.

**Reply by xPaw** [reply]
I think it's just straight up Steam giving bad servers, today I tried reconnecting to the same server in 15 second intervals and it worked after 5 tries. Even opening it in browser just gave connection reset.

**Reply by Shadowtail117** [reply]
Is there any way to get "good" servers reliably? My production build is hosted on the East Coast, whereas I'm in the Midwest, and neither have been able to reliably connect at all no matter how many times I attempt. My debug build got in once at random a few days ago but nothing since. At this point, I would accept connecting to a server halfway across the world if it meant being able to get in since my bot is effectively useless at the moment.

**Reply by ChrisN944** [reply]
> Is there any way to get "good" servers reliably? My production build is hosted on the East Coast, whereas I'm in the Midwest, and neither have been able to reliably connect at all no matter how many times I attempt. My debug build got in once at random a few days ago but nothing since. At this point, I would accept connecting to a server halfway across the world if it meant being able to get in since my bot is effectively useless at the moment.

Same. I would take anything. I'm on the final leg of development but can't fix the last remaining bugs until I get reliable access to Steam. Really disappointing.

**Comment by Richard-Whalley** [comment]
I would suggest you need to reduce your interval. I had the same issue, but when I reduced from 5 seconds to 250ms it actually connected. On Oct 18, 2024, at 9:54 PM, ChrisN944 ***@***.***> wrote:﻿

Is there any way to get "good" servers reliably? My production build is hosted on the East Coast, whereas I'm in the Midwest, and neither have been able to reliably connect at all no matter how many times I attempt. My debug build got in once at random a few days ago but nothing since. At this point, I would accept connecting to a server halfway across the world if it meant being able to get in since my bot is effectively useless at the moment.

Same. I would take anything. I'm on the final leg of development but can't fix the last remaining bugs until I get reliable access to Steam. Really disappointing.

**Reply by ChrisN944** [reply]
I've been running at 500 ms. I'll try reducing it.

**Reply by ChrisN944** [reply]
I was doing 3 retries with a 500 ms delay. That didn't work. I tried a 100 ms delay with 10 retries. That didn't work. I tried using a 250 ms delay with 10 retries, and that seems to do the trick... 

Hopefully Steam patches this soon, but thanks for the advice. I can at least continue development now. Much appreciated.

**Reply by Shadowtail117** [reply]
I haven't had any such luck sadly, even after changing it to a 250ms delay. Interestingly, sometimes the callback from `onDisconnected` is saying that the disconnect was user-initiated some of the time, though I never receive the onConnected callback, and the `isConnected` property of the client is still always false.
